staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,428 @@
+							t.Errorf("expected non-nil error")
+						}
+						if event != nil {
+							t.Errorf("expected nil event, got %v", *event)
+						}
+						// Restore startIndex.
+						wc.startIndex = originalCacheStartIndex
+					}
+				}
+
+				// Check if the state of the interval buffer is as expected.
+				// The interval buffer can be empty either when received is
+				// either a multiple of bufferSize (after one complete fill)
+				// or when received is equal to the number of expected events.
+				// The latter happens when partial filling occurs and no more
+				// events are left post the partial fill.
+				if wci.buffer.isEmpty() != (i%bufferSize == 0 || i == numExpectedEvents) {
+					t.Error("expected empty interval buffer")
+					return
+				}
+
+				event, err := wci.Next()
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+					return
+				}
+
+				expectedIndex := (c.intervalStartIndex + i) % wc.capacity
+				expectedEvent := wc.cache[expectedIndex]
+				if err := verifyEvent(true, event, expectedEvent); err != nil {
+					t.Error(err)
+				}
+			}
+			event, err := wci.Next()
+			ok := err != nil
+			if err := verifyNoEvent(ok, event); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestCacheIntervalNextFromStore(t *testing.T) {
+	getAttrsFunc := func(obj runtime.Object) (labels.Set, fields.Set, error) {
+		pod, ok := obj.(*v1.Pod)
+		if !ok {
+			return nil, nil, fmt.Errorf("not a pod")
+		}
+		return labels.Set(pod.Labels), fields.Set{"spec.nodeName": pod.Spec.NodeName}, nil
+	}
+	const numEvents = 50
+	store := cache.NewIndexer(storeElementKey, storeElementIndexers(nil))
+	events := make(map[string]*watchCacheEvent)
+	var rv uint64 = 1 // arbitrary number; rv till which the watch cache has progressed.
+
+	for i := 0; i < numEvents; i++ {
+		elem := makeTestStoreElement(
+			makeTestPod(
+				fmt.Sprintf("pod%d", i+1),
+				uint64(i+1),
+			),
+		)
+		objLabels, objFields, err := getAttrsFunc(elem.Object)
+		if err != nil {
+			t.Error(err)
+		}
+		events[elem.Key] = &watchCacheEvent{
+			Type:            watch.Added,
+			Object:          elem.Object,
+			ObjLabels:       objLabels,
+			ObjFields:       objFields,
+			Key:             elem.Key,
+			ResourceVersion: rv,
+		}
+		store.Add(elem)
+	}
+
+	wci, err := newCacheIntervalFromStore(rv, store, getAttrsFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < numEvents; i++ {
+		// The interval buffer can never be empty unless
+		// all elements obtained through List() have been
+		// returned.
+		if wci.buffer.isEmpty() && i != numEvents {
+			t.Fatal("expected non-empty interval buffer")
+		}
+		event, err := wci.Next()
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		if event == nil {
+			if i < numEvents {
+				t.Error("unexpected nil event")
+			}
+			break
+		}
+		expectedEvent, ok := events[event.Key]
+		if !ok {
+			t.Fatalf("event with key %s not found", event.Key)
+		}
+		if !reflect.DeepEqual(event, expectedEvent) {
```

Comment:
One more thing we want to test is whether we didn't return the same key twice - should be easy to add.

Reply:
@wojtek-t quick question - in either case of returning from watch cache or the store, I'm assuming this is to explicitly state and assert that duplicates are not happening? because the `reflect.DeepEqual` would take care of the duplicate return case as well right?

Reply:
OK fair - let's ignore this comment :)